deku-core/src/Deku/Core.purs

```diff
@@ -232,316 +157,170 @@ unsafeAttribute = Attribute
+-- |   empty. 
+-- | - after: This is more straightforward: for `Node`s it's simply after the `Node` and for `ParentNode` its before the
+-- |   first child `Node`.
+type BeamRegion =
+  EffectFn3 Anchor Anchor Anchor Unit
+
+-- | Generates a place for portals to render into that will not be displayed immediatly.
+type BufferPortal =
+  Effect DekuParent
 
 -- | This is the interpreter that any Deku backend creator needs to impelement.
--- | Three interpreters are included with Deku: SPA, SSR, and hydrated SSR.
--- | As an example, if you want to create a nullary interpreter that
--- | spits out `unit`, you can set everything to `mempty`.
-
+-- | Three interpreters are included with Deku: SPA.
+-- , SSR, and hydrated SSR.
 newtype DOMInterpret = DOMInterpret
   { makeElement :: MakeElement
   , setProp :: SetProp
   , setCb :: SetCb
-  , setDelegateCb :: SetDelegateCb
   , unsetAttribute :: UnsetAttribute
-  , attributeElementParent :: AttributeElementParent
-  , attributeDynParentForElement :: AttributeDynParentForElement
-  , sendToPosForElement :: SendToPosForElement
-  , removeForElement :: RemoveForElement
-  --
-  , makeOpenBeacon :: MakeBeacon
-  , makeCloseBeacon :: MakeBeacon
-  , attributeBeaconParent :: AttributeBeaconParent
-  , attributeDynParentForBeacons :: AttributeDynParentForBeacons
-  , attributeBeaconFullRangeParent :: AttributeBeaconFullRangeParent
-  , attributeDynParentForBeaconFullRange :: AttributeDynParentForBeaconFullRange
-  , sendToPosForDyn :: SendToPosForDyn
-  , removeForDyn :: RemoveForDyn
+  , attachElement :: AttachElement
+  , removeElement :: RemoveElement
   --
   , makeText :: MakeText
   , setText :: SetText
-  , attributeTextParent :: AttributeTextParent
-  , attributeDynParentForText :: AttributeDynParentForText
-  , sendToPosForText :: SendToPosForText
-  , removeForText :: RemoveForText
-  --
-  , toTemplate :: ToTemplate
-  , cloneElement :: CloneElement
-  , cloneTemplate :: CloneTemplate
-  , templateContent :: TemplateContent
+  , attachText :: AttachText
+  , removeText :: RemoveText
+  -- 
+  , bufferPortal :: BufferPortal
+  , beamRegion :: BeamRegion
   }
 
 derive instance Newtype DOMInterpret _
 
-getLifecycle
-  :: forall r
-   . Maybe
-       { end :: DekuBeacon
-       , lifecycle :: Maybe (Poll DekuDynamic)
-       , start :: DekuBeacon
-       , lucky :: Ref Boolean
-       | r
-       }
-  -> Maybe
-       { e :: DekuBeacon
-       , l :: Poll DekuDynamic
-       , s :: DekuBeacon
-       , lucky :: Ref Boolean
-       }
-getLifecycle mb = do
-  m <- mb
-  l <- m.lifecycle
-  pure { l, s: m.start, e: m.end, lucky: m.lucky }
-
-thunker :: STArray.STArray Global (Effect Unit) -> Effect Unit
-thunker unsubs = do
-  unsubsX <- liftST $ STArray.unsafeFreeze unsubs
-  runEffectFn1 fastForeachThunkE unsubsX
-
-runListener
+collectUnsubs :: EffectFn1 PSR (STArray.STArray Global (Effect Unit))
+collectUnsubs = mkEffectFn1 \(PSR psr) -> do
+  unsubs <- liftST $ STArray.new
```

Comment:
Maybe use this: <https://pursuit.purescript.org/packages/purescript-arrays/7.3.0/docs/Data.Array.ST#v:thaw>. That way you wouldn't need the null check and this could be a one-liner.

Reply:
We could also just hide the implementation of unsubs and just provide a `defer` and `drop/dispose?` effect in `PSR`. Under the hood we can then just use a mutable array and avoid copying altogether.

Reply:
Even better, great idea!